homeassistant/components/onvif/camera.py

```diff
@@ -108,56 +136,112 @@ def __init__(self, hass, config):
-
-        self._ptz_service = \
-            onvif.ONVIFService('http://{}:{}/onvif/device_service'.format(
-                self._host, self._port),
-                               self._username, self._password,
-                               '{}/wsdl/ptz.wsdl'.format(os.path.dirname(
-                                   onvif.__file__)))
-
-    def obtain_input_uri(self):
+
+        _LOGGER.debug("Setting up the ONVIF camera device @ '%s:%s'",
+                      self._host,
+                      self._port)
+
+        self._camera = ONVIFCamera(self._host,
+                                   self._port,
+                                   self._username,
+                                   self._password,
+                                   '{}/wsdl/'
+                                   .format(os.path.dirname(onvif.__file__)))
+
+    async def async_initialize(self):
+        """
+        Initialize the camera.
+
+        Initializes the camera by obtaining the input uri and connecting to
+        the camera. Also retrieves the ONVIF profiles.
+        """
+        _LOGGER.debug("Setting up the ONVIF device management service")
+
+        self._devicemgmt = self._camera.create_devicemgmt_service()
+
+        _LOGGER.debug("Retrieving current camera date/time")
+
+        system_date = dt.datetime.utcnow()
+        cdate = self._devicemgmt.GetSystemDateAndTime().UTCDateTime
+        cam_date = dt.datetime(cdate.Date.Year, cdate.Date.Month,
+                               cdate.Date.Day, cdate.Time.Hour,
+                               cdate.Time.Minute, cdate.Time.Second)
+
+        _LOGGER.debug("Camera date/time: %s",
+                      cam_date)
+
+        _LOGGER.debug("System date/time: %s",
+                      system_date)
+
+        dt_diff = cam_date - system_date
+        dt_diff_seconds = dt_diff.total_seconds()
+
+        if dt_diff_seconds > 5:
+            _LOGGER.warning("The date/time on the camera is '%s', "
+                            "which is different from the system '%s', "
+                            "this could lead to authentication issues",
+                            cam_date,
+                            system_date)
+
+        _LOGGER.debug("Obtaining input uri")
+
+        await self.async_obtain_input_uri()
+
+        _LOGGER.debug("Setting up the ONVIF PTZ service")
+
+        self._ptz_service = self._camera.create_ptz_service()
+
+        _LOGGER.debug("Completed set up of the ONVIF camera component")
+
+    async def async_obtain_input_uri(self):
         """Set the input uri for the camera."""
         from onvif import exceptions
+
         _LOGGER.debug("Connecting with ONVIF Camera: %s on port %s",
                       self._host, self._port)
 
         try:
-            profiles = self._media_service.GetProfiles()
+            _LOGGER.debug("Retrieving profiles")
+
+            media_service = self._camera.create_media_service()
+
+            profiles = media_service.GetProfiles()
```

Comment:
Should this also be awaited since it's I/O?  Not 100% sure how `zeep[async]` handles these requests.

Reply:
Found this example:

https://github.com/mvantellingen/python-zeep/blob/master/examples/async_client.py#L30

So I set up the loop, transport, etc, just not sure if this can be awaited or whether I have to use that future stuff.

Reply:
Found some docs, apparently they can be awaited as usual:

https://docs.python.org/3/library/asyncio-task.html

Reply:
Unfortunately hitting this:

https://github.com/FalkTannhaeuser/python-onvif-zeep/blob/zeep/onvif/client.py#L233

The onvif-zeep lib doesn't seem to support async out of the box. Maybe I can work around it.

Reply:
See https://github.com/FalkTannhaeuser/python-onvif-zeep/issues/43